### PR TITLE
[Optimizer] Greedy solution for join nodes in L1 Interleaved policy

### PR DESCRIPTION
**Pull request by fbajraktariTT**
This PR introduces new MemoryLayoutAnalysis policy as an alternative to the DFSharding policy with the goal of greedily solving join nodes. This means that if an op is a join node (node with multiple operands, e.g. Add, Matmul,...) we should consider placing multiple op's operands in the L1 memory and limit ourselves to only one operand being placed in L1 memory as is it the case with DFSharding. For all the combination of operands, we pick the one that maximizes L1 usage (by maximising L1 usage we try to have as few DRAM spills as possible) and minimizes required L1 usage (minimal amount of L1 memory neccesseary in order to reach desired memory configuration). Furthermore, it should be emphasized that this may not lead to globally optimal solution because we pick the optimal config at moment of analysing the op.

### CODE REVIEW

File: lib/Scheduler/Scheduler.cpp [hunk 1]
```diff
@@ -22,19 +22,19 @@ bool isTTIROp(mlir::Operation *op) {
 }
 
 bool isTTShedulableOp(mlir::Operation *op) {
-  return isTTNNOp(op) || isTTIROp(op);
+  return (isTTNNOp(op) || isTTIROp(op)) && (not isa<func::ReturnOp>(op));
```
Comment (nobradovictt) [inline lib/Scheduler/Scheduler.cpp:25]: image

I think it would be better to check traits and see if its "return-like".

**Review by odjuricicTT — COMMENTED** [review]
As discussed offline, because it is a big change, let's make a few architectural adjustments first and then review the rest.

File: include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h [hunk 2]
```diff
@@ -8,10 +8,31 @@
...
+public:
+  struct OpMemSpec {
+    TTNNLayoutAttr layout;
+    // Minimum L1 memory usage required for scheduling the op
+    // given the layouts of all the ops that are already scheduled.
+    //
+    uint64_t requiredL1Usage;
+  };
+
+  struct Config {
```
Comment (odjuricicTT) [inline include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h:25]: What is this config for? Can you use a more descriptive name and add comments?

File: include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h [hunk 3]
```diff
@@ -23,6 +44,58 @@ class L1InterleavedPolicy : public MemoryLayoutAnalysisPolicy {
...
+    // Figure out this const based on exec data, but will be replaced
+    // with API.
+    //
+    constexpr float tensorL1UsageCap = 0.75;
+    return tensorL1UsageCap * usableL1CacheSize;
+  }
+
+  // Get optimal op's config base on the outputs of its operands.
+  Config getOptimalConfig(Operation *baseOp,
+                          llvm::DenseMap<Operation *, L1Usage> &opsL1Usage);
```
Comment (odjuricicTT) [inline include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h:71]: Are opsL1Usage operands of baseOp? If so, maybe a better name would be operandL1Usage?

File: include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h [hunk 4]
```diff
@@ -23,6 +44,58 @@ class L1InterleavedPolicy : public MemoryLayoutAnalysisPolicy {
...
+
+  // Precedence schedule map for each operation.
+  llvm::DenseMap<Operation *, llvm::SmallVector<Operation *>> precedenceMap;
```
Comment (odjuricicTT) [inline include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h:74]: Add additional comment. "Contains the order in which operands need to be executed for each op"?

File: include/ttmlir/Scheduler/Scheduler.h [hunk 5]
```diff
@@ -23,6 +23,9 @@ class Scheduler {
   // Method to get the next set of schedulable operations
   llvm::SmallVector<mlir::Operation *> getScheduleableOps();
 
+  // Method to check if an operation is a TT schedulable
+  bool isTTShedulableOp(mlir::Operation *op);
```
Comment (odjuricicTT) [inline include/ttmlir/Scheduler/Scheduler.h:27]: Method to check if an operation is a TTIR or TTNN op

File: lib/Dialect/TTNN/Analysis/MemoryLayoutAnalysis.cpp [hunk 6]
```diff
@@ -35,14 +36,15 @@ filterShardedOnly(const llvm::DenseMap<Operation *, std::vector<TTNNLayoutAttr>>
...
-filterL1InterleavedOnly(
+filterDRAMAndL1Interleaved(
     const llvm::DenseMap<Operation *, std::vector<TTNNLayoutAttr>>
         &legalLayouts) {
   llvm::DenseMap<Operation *, std::vector<TTNNLayoutAttr>> l1InterleavedLayouts;
   for (const auto &opLayouts : legalLayouts) {
     std::vector<TTNNLayoutAttr> opL1InterleavedLayouts;
     for (const auto &layout : opLayouts.second) {
-      if (layout.hasInterleavedL1TensorMemoryLayout()) {
+      if (layout.getBufferType() == BufferType::DRAM ||
```
Comment (odjuricicTT) [inline lib/Dialect/TTNN/Analysis/MemoryLayoutAnalysis.cpp:46]: Can you use hasDRAMTensorMemoryLayout?

File: lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp [hunk 7]
```diff
@@ -34,6 +39,16 @@ bool TTNNLayoutAttr::isTiled() const {
   return ::mlir::isa<::mlir::tt::TileType>(getElementType());
 }
 
+// Check if the tensor memory layout is in L1 memory
+bool TTNNLayoutAttr::hasL1TensorMemoryLayout() const {
```
Comment (odjuricicTT) [inline lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp:43]: We do have problems with conflicting naming with tensor related stuff in general. But doesn't tensor memory layout refer to interleaved / sharded, does it make more sense to call this hasL1BufferType?

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 8]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
-void L1InterleavedPolicy::run() {
-  rootOp->walk([&](func::FuncOp func) {
-    DeviceAttr deviceAttr = getCurrentScopeDevice(func);
-    mlir::tt::scheduler::Scheduler scheduler(&func);
-    llvm::SmallVector<mlir::Operation *> scheduleableOps;
-    llvm::DenseMap<Operation *, TTNNLayoutAttr> selectedOpLayout;
-    Operation *currentOp = nullptr;
+L1InterleavedPolicy::Config L1InterleavedPolicy::getOptimalConfig(
+    Operation *baseOp, llvm::DenseMap<Operation *, L1Usage> &opsL1Usage) {
+  uint64_t n, i, j, currentMask;
```
Comment (odjuricicTT) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:33]: Don't name variables with single like this if they are not simple itterators.

Comment (odjuricicTT) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:33]: Seems like `i` can just be `currentMask` etc

**Review by odjuricicTT — COMMENTED** [review]
Left a few comments, mostly about readability. As @nobradovic mentioned, the PR title should try to explain what the change is in a way so that people not familiar with the work can uderstand. Maybe something like: "[Optimizer] Solve join nodes in L1 interleaved policy"

File: include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h (hunk 3 above)
Comment (fbajraktariTT) [reply include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h:71]: That is not true. The `opsL1Usage` contains all the information about the execution of the `baseOp`. So, `L1Usage`s of both `baseOp` and its operands is stored in the `opsL1Usage` object. The reason for this is that the algorithm for determining the optimal l1 usage of the op does not make distinction between 'baseOp' and its operands. It just creates the configuration of L1 layouts so that the L1 memory is maximally used.

File: include/ttmlir/Scheduler/Scheduler.h (hunk 5 above)
Comment (fbajraktariTT) [reply include/ttmlir/Scheduler/Scheduler.h:27]: Can you explain further what do you mean by this comment? The idea here was to expose the `isTTShedulableOp(mlir::Operation *op)` as a public method and, therefore, in the `L1InrterleavedPolicy` discard non-analyzable ops (e.g. `EmptyOp`s).

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 9]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
+  uint64_t currentL1Usage, optimalL1Usage;
+  llvm::DenseMap<Operation *, TTNNLayoutAttr> optimalLayouts;
+  llvm::SmallVector<Operation *> optimalPrecedence;
+
+  optimalL1Usage = 0;
+  n = opsL1Usage.size();
+  for (i = 0; i < (1 << n); i++) {
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp]: Whats this? No comments and it looks like it can overflow.

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 10]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
-    // TODO(fbajraktari): Add algorithm description. Currently, the algorithm
-    // is the same as for DFSharding policy, but works only for L1 interleaved.
+    // Calculate the L1 usage of the current configuration.
     //
-    l1ChainConfigs->push_back(L1ChainConfig());
-    while (scheduler.hasUnscheduledOps()) {
-      scheduleableOps = scheduler.getScheduleableOps();
+    currentL1Usage = 0;
+    j = 1;
+    currentMask = i;
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp]: Can be written much cleaner and safer using std::bitset.

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 11]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
 }
 
-void L1InterleavedPolicy::run() {
-  rootOp->walk([&](func::FuncOp func) {
-    DeviceAttr deviceAttr = getCurrentScopeDevice(func);
-    mlir::tt::scheduler::Scheduler scheduler(&func);
-    llvm::SmallVector<mlir::Operation *> scheduleableOps;
-    llvm::DenseMap<Operation *, TTNNLayoutAttr> selectedOpLayout;
-    Operation *currentOp = nullptr;
+L1InterleavedPolicy::Config L1InterleavedPolicy::getOptimalConfig(
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp]: Comment what is the purpose of this function and how it works.

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp (hunk 9 above)
Comment (nobradovictt) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp]: I would also like to see some estimates how will this work in worst case.

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 12]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
-              break;
-            }
+        scheduler.scheduleOp(op);
+
+        // Find optimal configuration for the op.
+        //
+        llvm::DenseMap<Operation *, L1Usage> opsL1Usage;
+        llvm::SmallVector<Operation *> opsPrecedence;
+
+        // Generate optimal configuration for the current op based on the
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:170]: What is the case when this is false?

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 13]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
+        //
+        if (isAnalyzable(op)) {
+
+          // Create the OpMemSpec.
+          //
+          OpMemSpec OpMemSpec;
+          assert(hasDRAMLayout(op));
+          OpMemSpec.layout = getDRAMLayout(op);
+          OpMemSpec.requiredL1Usage = 0;
+          OpMemSpecMap[op] = std::move(OpMemSpec);
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:181]: Pointless std::move

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 14]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
+
+          if (op->hasOneUse() && hasL1InterleavedLayout(op)) {
+            L1Usage l1Usage;
+            l1Usage.outputL1Usage =
+                getOpOutputL1Usage(op, getL1InterleavedLayout(op), deviceAttr);
+            l1Usage.requiredL1Usage = 0;
+            opsL1Usage[op] = std::move(l1Usage);
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:188]: Another pointless std::move

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 15]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
-              // Update currentOp pointer.
-              //
-              currentOp = nextOp;
-              continue;
+            if (operandOpLayout.hasInterleavedL1TensorMemoryLayout()) {
+              L1Usage l1Usage;
+              l1Usage.outputL1Usage =
+                  getOpOutputL1Usage(operandOp, operandOpLayout, deviceAttr);
+              l1Usage.requiredL1Usage = OpMemSpecMap[operandOp].requiredL1Usage;
+              opsL1Usage[operandOp] = std::move(l1Usage);
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:214]: Pointless move

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 16]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
     //
-    (*schedule)[func] = scheduler.getSchedule();
+    l1ChainConfigs->push_back(L1ChainConfig());
+    llvm::DenseMap<Operation *, TTNNLayoutAttr> selectedOpLayout;
+    for (auto &OpMemSpec : OpMemSpecMap) {
+      OpL1MemSpec opL1MemSpec;
+      opL1MemSpec.op = OpMemSpec.first;
+      opL1MemSpec.tensorSplitFactor = 1;
+      selectedOpLayout[OpMemSpec.first] = OpMemSpec.second.layout;
+      l1ChainConfigs->back().addOpL1MemSpec(std::move(opL1MemSpec));
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:299]: Pointless move.

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp [hunk 17]
```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
...
+    }
+    l1ChainConfigs->back().build();
+    l1ChainConfigs->back().resolve();
+    std::unordered_set<Edge> memReconfigEdges;
```
[the remark this answers is not in the document]
Comment (fbajraktariTT) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:303]: It depends on the op and there may be a scenario where we need reconfigEdges. For example, if we greedily decide that the output of an op's operand should be in DRAM and there is enough available L1 memory this operand then we can insert a ToLayout op on that edge so we can gain on the performance. I assume this may be true depending on the op.

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp (hunk 9 above)
Comment (fbajraktariTT) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp]: In what sense do you mean worst case?

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp (hunk 12 above)
Comment (fbajraktariTT) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:170]: There are two ways that the op is not analysable: 1) Op does not produce any output (e.g. `ttnn.get_device()`) and 2) all op's legal layouts are filtered out in `MemoryLayoutAnalysis` (this is true for the last `ToLayout` op that converts tensor back to `System` memory and therefore all its only legal layout will be filtered out since it not DRAM nor L1 buffer type).

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp (hunk 16 above)
Comment (fbajraktariTT) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:299]: You added that `addOpL1MemSpec` requires rvalue reference.

Comment (nobradovictt) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:302]: Please fix that as well, I might have prototyped with OpL1MemSpec having vector or thinking move would help for Layout object.

File: include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h [hunk 18]
```diff
@@ -23,6 +56,69 @@ class L1InterleavedPolicy : public MemoryLayoutAnalysisPolicy {
...
+  size_t getAvailableL1CacheSize() const {
+    // Figure out this const based on exec data, but will be replaced
+    // with API.
+    //
+    constexpr float tensorL1UsageCap = 0.75;
+    return tensorL1UsageCap * usableL1CacheSize;
+  }
+
+  /**
+   * Get the greedily optimal OpConfig for the provided base operation
```
Comment (nobradovictt) [inline include/ttmlir/Dialect/TTNN/Analysis/L1InterleavedPolicy.h:82]: This is a strange bittersweet combination of words.

**Comment by nobradovictt** [comment]
Change has no description.

File: include/ttmlir/Scheduler/Scheduler.h (hunk 5 above)
Comment (odjuricicTT) [reply include/ttmlir/Scheduler/Scheduler.h:27]: Sorry, i wasn't clear here, it was just a nitpick. I meant that "Method to check if an operation is a TTIR or TTNN op" might be a better comment for the method as it provides more details on what the method does. With the current name and comment i had to got to the implementation to actually understand what it checks.

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp (hunk 16 above)
Comment (odjuricicTT) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp:302]: Just to clarify, @nobradovictt you mean that addOpL1MemSpec should just take OpL1MemSpec by value?

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp (hunk 9 above)
Comment (fbajraktariTT) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp]: If an op has `n` operands, the size of `opsL1Usage` will not be larger than `n+1`. Since our ops have at most 2 operands (without including `EmptyOp` as an DPS op's argument), `n` will be at most 3.

Comment (nobradovictt) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp]: Add this is comment on assignment to n. Also think about naming it better(than n), like numOperandsBound or something similar.

File: lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp (hunk 16 above)
Comment (nobradovictt) [reply lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp]: Yeah its a simple structure if TTNNLayoutAttr should be copied by value and I think it should based on MLIR docs I read.
image
https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/#:~:text=The%20C%2B%2B%20Attribute%20and%20Type,%2Dpointer%20or%20by%2Dreference.

File: include/ttmlir/Dialect/TTNN/IR/TTNNOpsAttrs.td [hunk 19]
```diff
@@ -148,6 +148,9 @@ def TTNN_TTNNLayoutAttr: TTNN_Attr<"TTNNLayout", "ttnn_layout"> {
       bool hasShardedTensorMemoryLayout() const;
       bool hasShardedL1TensorMemoryLayout() const;
       bool hasInterleavedL1TensorMemoryLayout() const;
+      bool hasInterleavedDRAMTensorMemoryLayout() const;
+      bool hasL1BufferTypeLayout() const;
+      bool hasDRAMBufferTypeLayout() const;
```
[the remark this answers is not in the document]
Comment (fbajraktariTT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOpsAttrs.td:153]: Maybe, `hasInterleavedDRAMTensorMemoryLayout()` should stay as it is since `interleaved` is related to tensor layout.

File: lib/Dialect/TTNN/Analysis/CMakeLists.txt [hunk 20]
```diff
@@ -15,6 +15,6 @@ add_mlir_dialect_library(MLIRTTNNAnalysis
         MLIRTTNNPassesIncGen
         MLIRTTOpsIncGen
 
-        LINK_LIBS
+        LINK_LIBS PUBLIC
```
Comment (odjuricicTT) [inline lib/Dialect/TTNN/Analysis/CMakeLists.txt:18]: Just to give me context, why is this needed?

**Review by odjuricicTT — APPROVED** [review]
Looks good! Please add a description for the PR with a few sentances describing the PR contains. Do this for all future PRs.

File: lib/Dialect/TTNN/Analysis/CMakeLists.txt (hunk 20 above)
Comment (fbajraktariTT) [reply lib/Dialect/TTNN/Analysis/CMakeLists.txt:18]: My understanding is that is similar to c++ public inheritance. Let's say we want to link lib A into lib B and then lib B into C. If we assume that the type of linking is PUBLIC for both libs, the lib C should implicitly link the lib C. @mtopalovicTT

**Review by nobradovictt — APPROVED** [review]
